modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotDeleteSafeZoneCalc.java

```diff
@@ -103,22 +80,50 @@ public void run() {
                         continue;
                     }
 
-                    if (null != safeZoneByPublishedMessages) {
-                        safeZoneValue = safeZoneByPublishedMessages;
+                    /** calculate safe zone (minimum value of messageIDs published so far to the
+                     * cluster by each node)
+                     */
+                    long globalSafeZoneVal = Long.MAX_VALUE;
+                    for (String nodeID : nodesWithPublishedMessages) {
+
+                        long safeZoneValue = Long.MAX_VALUE;
+
+                        //get the maximum message id published by node so far
+                        Long safeZoneByPublishedMessages;
+                        try {
+                            safeZoneByPublishedMessages = SlotManagerClusterMode.getInstance()
+                                    .getLocalSafeZone(nodeID);
+
+                        } catch (AndesException e) {
+                            log.error("SlotDeleteSafeZoneCalc stopped due to failing to get last published id for node:" +
+                                    nodeID + ". Retrying after 15 seconds", e);
+                            try {
+                                Thread.sleep(TIME_TO_SLEEP_ON_ERROR);
```

Comment:
I've created a JIRA to change this implementation to a scheduled task.
Tracked by https://wso2.org/jira/browse/MB-1615